### PR TITLE
doc/rados: edit crush-map.rst (3 of x)

### PR DESCRIPTION
**Pull request by zdover23**
Edit doc/rados/operations/crush-map.rst.

https://tracker.ceph.com/issues/58485

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/rados/operations/crush-map.rst [hunk 1]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+form:
 
 .. prompt:: bash $
 
    ceph osd crush rule rm {rule-name}
 
-
 .. _crush-map-tunables:
 
 Tunables
```
Comment (zdover23) [inline doc/rados/operations/crush-map.rst:743]: A future PR (a very near-future PR perhaps as early as forty-eight hours from the time of this comment) will reorganize this tunable-based trip down memory lane into something that looks better in RST.

File: doc/rados/operations/crush-map.rst [hunk 2]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+   value is 0.
 
- * ``choose_total_tries``: Total number of attempts to choose an item.
-   Legacy value was 19, subsequent testing indicates that a value of
-   50 is more appropriate for typical clusters.  For extremely large
-   clusters, a larger value might be necessary.
+ * ``choose_total_tries``: Total number of attempts to choose an item.  The
+   legacy value is ``19``, but subsequent testing indicates that a value of
+   ``50`` is more appropriate for typical clusters. For extremely large
+   clusters, an even larger value might be necessary.
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:795]: Been there, on a cluster with only ~60 OSDs, on Luminous even.

File: doc/rados/operations/crush-map.rst [hunk 3]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
 
- * ``chooseleaf_descend_once``: Whether a recursive chooseleaf attempt
-   will retry, or only try once and allow the original placement to
-   retry.  Legacy default is 0, optimal value is 1.
+ * ``chooseleaf_descend_once``: Whether a recursive chooseleaf attempt will
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:797]: ``chooseleaf``?

File: doc/rados/operations/crush-map.rst [hunk 4]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
    populated with data.
 
 firefly (CRUSH_TUNABLES3)
 -------------------------
 
-The ``firefly`` tunable profile fixes a problem
-with ``chooseleaf`` CRUSH rule behavior that tends to result in PG
-mappings with too few results when too many OSDs have been marked out.
+The ``firefly`` tunable profile fixes a problem with ``chooseleaf`` CRUSH rule
+behavior. When this problem arises, PG mappings with not enough results can
```
Comment (anthonyeleven) [inline doc/rados/operations/crush-map.rst:811]: s/with not enough results/too few OSDs/

**Review by anthonyeleven — APPROVED** [review]

File: doc/rados/operations/crush-map.rst (hunk 3 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:797]: I don't know. @neha-ojha?

Comment (anthonyeleven) [reply doc/rados/operations/crush-map.rst:797]: I meant to double-backtick on like 797, markup only.

File: doc/rados/operations/crush-map.rst [hunk 5]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
 -------------------------------------
 
-For a replicated pool, the primary decision when creating the CRUSH
-rule is what the failure domain is going to be.  For example, if a
-failure domain of ``host`` is selected, then CRUSH will ensure that
-each replica of the data is stored on a unique host.  If ``rack``
-is selected, then each replica will be stored in a different rack.
-What failure domain you choose primarily depends on the size and
+When you create a CRUSH rule for a replicated pool, there is an
+important decision to make: you must select a failure domain. For
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:622]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 6]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+example, if you select a failure domain of ``host``, then CRUSH will
+ensure that each replica of the data is stored on a unique host.
+Alternatively, if you select a failure domain of ``rack``, then each
+replica of the data will be stored in a different rack. Your
+selection of failure domain should be guided by the size and
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:627]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 7]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
-It is also possible to create a rule that restricts data placement to
-a specific *class* of device.  By default, Ceph OSDs automatically
-classify themselves as either ``hdd`` or ``ssd``, depending on the
-underlying type of device being used.  These classes can also be
-customized.
+It is possible to create a rule that restricts data placement to
+a specific *class* of device. By default, Ceph OSDs automatically
+classify themselves as either ``hdd`` or ``ssd`` in accordance with
+the underlying type of device being used. These device classes can
+be customized.
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:640]: Accepted with minor modifications.

File: doc/rados/operations/crush-map.rst [hunk 8]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+   :Required: Yes
+   :Example: ``rbd-rule``
 
 ``root``
-
-:Description: The name of the node under which data should be placed.
-:Type: String
-:Required: Yes
-:Example: ``default``
+   :Description: The name of the node under which data is to be placed.
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:657]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 9]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
-For an erasure-coded (EC) pool, the same basic decisions need to be made:
-what is the failure domain, which node in the
-hierarchy will data be placed under (usually ``default``), and will
-placement be restricted to a specific device class.  Erasure code
-pools are created a bit differently, however, because they need to be
-constructed carefully based on the erasure code being used.  For this reason,
-you must include this information in the *erasure code profile*.  A CRUSH
-rule will then be created from that either explicitly or automatically when
-the profile is used to create a pool.
+For an erasure-coded pool, fundamentally similar decisions need to be made:
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:677]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 10]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
- * For large clusters, some small percentages of PGs map to fewer than
-   the desired number of OSDs.  This is more prevalent when there are
-   multiple hierarchy layers in use (e.g., ``row``, ``rack``, ``host``, ``osd``).
+ * For large clusters, a small percentage of PGs might map to fewer than the
+   desired number of OSDs. This is known to happen when there are multiple
+   hierarchy layers in use (for example,, ``row``, ``rack``, ``host``,
+   ``osd``).
 
- * When some OSDs are marked out, the data tends to get redistributed
+ * When one or more OSDs are marked ``out``, the data tends to be redistributed
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:781]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 11]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
-pools are created a bit differently, however, because they need to be
-constructed carefully based on the erasure code being used.  For this reason,
-you must include this information in the *erasure code profile*.  A CRUSH
-rule will then be created from that either explicitly or automatically when
-the profile is used to create a pool.
+For an erasure-coded pool, fundamentally similar decisions need to be made:
+what the failure domain is, which node in the hierarchy data will be placed
+under (usually ``default``), and whether placement is restricted to a specific
+device class. However, erasure-code pools are created in a different way: there
+is a need to construct them carefully with reference to the erasure code in
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:681]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 12]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
-the new version of CRUSH.  For this reason, we have created
-``profiles`` that are named after the Ceph version in which they were
-introduced.  For example, the ``firefly`` tunables are first supported
-by the Firefly release, and will not work with older (e.g., Dumpling)
-clients.  Once a given set of tunables are changed from the legacy
-default behavior, the ``ceph-mon`` and ``ceph-osd`` will prevent older
-clients who do not support the new CRUSH features from connecting to
-the cluster.
+The CRUSH algorithm that is used to calculate the placement of data has been
+improved over time. In order to support the change in behavior, we have
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:747]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 13]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+provided users with a set of tunable options that can determine whether the
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:748]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 14]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+legacy version or the improved version of CRUSH is to be used.
+
+In order to use newer tunables, both clients and servers must support the new
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:751]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 15]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+version of CRUSH. Because of this requirement, we have created ``profiles``
+that are named after the Ceph version in which they were introduced. For
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:753]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 16]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+example, the ``firefly`` tunables were first supported by the Firefly release
+and do not work with older clients (for example, clients running Dumpling).
+After a given set of tunables has been changed from the legacy default
+behavior, the ``ceph-mon`` and ``ceph-osd`` options will prevent older clients
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:757]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 17]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
 -------------------------
 
-The ``bobtail`` tunable profile fixes a few key misbehaviors:
+The ``bobtail`` tunable profile provides the following fixes:
 
- * For hierarchies with a small number of devices in the leaf buckets,
-   some PGs map to fewer than the desired number of replicas.  This
-   commonly happens for hierarchies with "host" nodes with a small
-   number (1-3) of OSDs nested beneath each one.
+ * For hierarchies with a small number of devices in the leaf buckets, some PGs
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:771]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 18]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+   might map to fewer than the desired number of replicas.  This is known to
+   happen in the case of hierarchies with "host" nodes that have a small number
+   of OSDs (1 to 3) nested beneath each host.
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:774]: Accepted.

File: doc/rados/operations/crush-map.rst [hunk 19]
```diff
@@ -618,285 +618,280 @@ To remove a weight set:
...
+   desired number of OSDs. This is known to happen when there are multiple
+   hierarchy layers in use (for example,, ``row``, ``rack``, ``host``,
+   ``osd``).
 
- * When some OSDs are marked out, the data tends to get redistributed
+ * When one or more OSDs are marked ``out``, the data tends to be redistributed
    to nearby OSDs instead of across the entire hierarchy.
 
-The new tunables are:
+The new tunables introduced by ``bobtail`` are as follows:
```
[the remark this answers is not in the document]
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:784]: Accepted. Good call.

File: doc/rados/operations/crush-map.rst (hunk 3 above)
Comment (zdover23) [reply doc/rados/operations/crush-map.rst:797]: Accepted.

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/52093 - Reef backport
https://github.com/ceph/ceph/pull/52094 - Quincy backport

**Pull request merged**